src/io_uring/src/lib.rs

```diff
@@ -225,3 +225,172 @@ impl IoUring {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    /// -------------------------------------
+    /// BEGIN PROPERTY BASED TESTING
+    use super::*;
+    use proptest::prelude::*;
+    use proptest::strategy::Strategy;
+    use proptest::test_runner::{Config, TestRunner};
+    use std::os::unix::fs::FileExt;
+    use utils::kernel_version::KernelVersion;
+    use utils::syscall::SyscallReturnCode;
+    use utils::tempfile::TempFile;
+    use vm_memory::{Bytes, MmapRegion, VolatileMemory};
+
+    fn drain_cqueue(ring: &mut IoUring) {
+        while let Some(entry) = ring.pop::<usize>().unwrap() {
+            assert!(entry.result().is_ok());
+
+            // Assert that there were no partial writes.
+            let count = entry.result().unwrap();
+            let user_data = entry.user_data() as u32;
+            assert_eq!(count, user_data);
+        }
+    }
+
+    fn arbitrary_operation(
+        file_len: u32,
+        mem_region: &MmapRegion,
+    ) -> impl Strategy<Value = Operation<u32>> {
+        let base_mem_ptr = mem_region.as_ptr();
+
+        (0u32..file_len)
+            .prop_flat_map(move |len| {
+                (
+                    // len
+                    Just(len),
+                    // offset
+                    (0u32..(file_len - len)),
+                    // mem region offset
+                    (0u32..(file_len - len)),
+                )
+            })
+            .prop_map(move |(len, off, mem_off)| {
+                let mem_addr = unsafe { base_mem_ptr.add(mem_off as usize) as usize };
+
+                let mut operation = Operation::write(0, mem_addr, len, off.into(), len);
+                // Make sure the operations are executed in-order, so that they are equivalent to
+                // their sync counterparts.
+                operation.ordered();
```

Comment:
Even if the order of the ops is not guaranteed, I was expecting for the final result to be consistent. Just like in the case of reordering CPU ops by the compiler or the CPU. I will run some tests.

Reply:
in my experiments the order was consistent, but I think it's better to enforce this in the proptest, so that we don't get false failures that are caused by reorders

Reply:
True. Agree.